src/datadog/telemetry/telemetry.cpp

```diff
@@ -1,176 +1,119 @@
+TelemetryProxy make_telemetry(const Ctor_param& init) {
+  if (!init.configuration.enabled) return NoopTelemetry{};
+  return Telemetry{init.configuration, init.logger,    init.client,
+                   init.metrics,       init.scheduler, init.agent_url,
+                   init.clock};
 }
 
-void Telemetry::log_warning(std::string message) {
-  tracer_telemetry_->log(std::move(message), LogLevel::WARNING);
+TelemetryProxy& instance(
+    const tracing::Optional<Ctor_param>& init = tracing::nullopt) {
+  static TelemetryProxy telemetry(make_telemetry(*init));
+  return telemetry;
 }
 
-void Telemetry::send_telemetry(StringView request_type, std::string payload) {
-  if (!config_.enabled) return;
-  auto set_telemetry_headers = [request_type, payload_size = payload.size(),
-                                debug_enabled = tracer_telemetry_->debug(),
-                                tracer_signature =
-                                    &tracer_signature_](DictWriter& headers) {
-    /*
-      TODO:
-        Datadog-Container-ID
-    */
-    headers.set("Content-Type", "application/json");
-    headers.set("Content-Length", std::to_string(payload_size));
-    headers.set("DD-Telemetry-API-Version", "v2");
-    headers.set("DD-Client-Library-Language", "cpp");
-    headers.set("DD-Client-Library-Version", tracer_signature->library_version);
-    headers.set("DD-Telemetry-Request-Type", request_type);
-
-    if (debug_enabled) {
-      headers.set("DD-Telemetry-Debug-Enabled", "true");
-    }
-  };
-
-  // TODO(@dmehala): make `clock::instance()` a singleton
-  auto post_result = http_client_->post(
-      telemetry_endpoint_, set_telemetry_headers, std::move(payload),
-      telemetry_on_response_, telemetry_on_error_,
-      tracing::default_clock().tick + 5s);
-  if (auto* error = post_result.if_error()) {
-    logger_->log_error(
-        error->with_prefix("Unexpected error submitting telemetry event: "));
-  }
+void init(FinalizedConfiguration configuration,
+          std::shared_ptr<tracing::Logger> logger,
+          std::shared_ptr<tracing::HTTPClient> client,
+          std::vector<std::shared_ptr<Metric>> metrics,
+          std::shared_ptr<tracing::EventScheduler> event_scheduler,
+          tracing::HTTPClient::URL agent_url, tracing::Clock clock) {
+  instance(Ctor_param{configuration, logger, client, metrics, event_scheduler,
+                      agent_url, clock});
 }
 
-void Telemetry::send_app_started(
-    const std::unordered_map<ConfigName, ConfigMetadata>& config_metadata) {
-  send_telemetry("app-started",
-                 tracer_telemetry_->app_started(config_metadata));
+void send_app_started(const std::unordered_map<tracing::ConfigName,
+                                               tracing::ConfigMetadata>& conf) {
+  std::visit(
+      details::Overload{
+          [&](Telemetry& telemetry) { telemetry.send_app_started(conf); },
+          [](NoopTelemetry) {},
```

Comment:
```suggestion
          [](auto&&) {},
```

Reply:
That's on purpose. While `auto&&` acts as a sink, I prefer the explicitness of the implementation. It reduces the risk of errors if someone, for any reason, decides to add a new value to the variant.

Reply:
Do we want to do the same in the below implementation ? Or the goal is to show the two versions ?

Reply:
Good catch, that's a mistake on my end